### PR TITLE
Server part of OpenThread CoAP example + some improvements in the client part

### PR DESCRIPTION
**Pull request by lmaciejonczyk**

### CODE REVIEW

**Comment by lmaciejonczyk** [comment]
There is needed a one additional change in the file: modules/lib/openthread/zephyr/CMakeLists.txt which is part of zephyr project. It is needed to enable OT CoAP api.

`if(CONFIG_OPENTHREAD_COAP)
  set(OT_COAP ON CACHE BOOL "Enable CoAP api")
endif()`

File: samples/net/openthread/coap_client/CMakeLists.txt [hunk 1]
```diff
@@ -7,14 +7,18 @@ cmake_minimum_required(VERSION 3.13.1)
 
 set(NRF_SUPPORTED_BOARDS
 	nrf52840_pca10056
+	nrf52833_pca10100
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_client/CMakeLists.txt:10]: cool

File: samples/net/openthread/coap_client/src/coap_client.c [hunk 2]
```diff
@@ -6,56 +6,44 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(coap_client, CONFIG_COAP_CLIENT_LOG_LEVEL);
 
-#include <logging/log_ctrl.h>
-#include <errno.h>
+#include "coap_utils.h"
+
+#include <coap_server_client_interface.h>
+
 #include <zephyr.h>
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_client/src/coap_client.c]: nit: keep zephyr includes above app includes

File: samples/net/openthread/coap_client/src/coap_utils.c [hunk 3]
```diff
@@ -30,7 +29,8 @@ const static int nfds = 1;
 static struct pollfd fds;
 static struct coap_reply replies[COAP_MAX_REPLIES];
 
-static void coap_open_socket(int *sock) {
+static void coap_open_socket(int *sock)
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_client/src/coap_utils.c]: why not return socket?

File: samples/net/openthread/coap_interface/coap_server_client_interface.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+typedef enum {
+	THREAD_COAP_UTILS_LIGHT_CMD_OFF = '0',
+	THREAD_COAP_UTILS_LIGHT_CMD_ON = '1',
+	THREAD_COAP_UTILS_LIGHT_CMD_TOGGLE = '2'
+} light_command;
+
+#define PROVISIONING_URI_PATH "provisioning"
+#define LIGHT_URI_PATH "light"
+
+#endif
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_interface/coap_server_client_interface.h]: newline

File: samples/net/openthread/coap_interface/coap_server_client_interface.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+#define __COAP_SERVER_CLIENT_INTRFACE_H__
+
+#define COAP_PORT 5683
+
+/**@brief Enumeration describing light commands. */
+typedef enum {
+	THREAD_COAP_UTILS_LIGHT_CMD_OFF = '0',
+	THREAD_COAP_UTILS_LIGHT_CMD_ON = '1',
+	THREAD_COAP_UTILS_LIGHT_CMD_TOGGLE = '2'
+} light_command;
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_interface/coap_server_client_interface.h]: Add _t suffix as this is typedef. At least that's what others do.

File: samples/net/openthread/coap_server/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#
+cmake_minimum_required(VERSION 3.13.1)
+
+set(NRF_SUPPORTED_BOARDS
+	nrf52840_pca10056
+	nrf52833_pca10100
+)
+
+# IMPORTANT: this must come before including boilerplate.cmake.
+set(KCONFIG_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/Kconfig)
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/CMakeLists.txt]: Why is this needed?

File: samples/net/openthread/coap_server/README.rst [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+Overview
+********
+
+
+Requirements
+************
+
+* One of the following development boards:
+
+  * |nRF52840|
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/README.rst]: Please add nrf52833_pca10100 as you mentioned it is supported in yaml and cmake, also update client rst. BTW, not sure it is merged to NCS already but there is no nrf52833_pca10100 and nrf52840_pca10056 anymore - they were renamed.
I think |nRF52840| link points to the SOC not BOARD, but can't verify that.

File: samples/net/openthread/coap_server/src/coap_server.c [hunk 8]
```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <logging/log.h>
+LOG_MODULE_REGISTER(coap_server, CONFIG_COAP_SERVER_LOG_LEVEL);
+
+#include "ot_coap_utils.h"
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/coap_server.c:11]: nit: move this below zephyr includes

File: samples/net/openthread/coap_server/src/coap_server.c [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+		LOG_INF("Provisioning deactivated");
+	}
+}
+
+static void on_provisioning_timer_expiry(struct k_timer *timer_id)
+{
+	deactivate_provisionig();
+}
+
+static void on_let_timer_expiry(struct k_timer *timer_id)
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/coap_server.c]: typo: I think it was supposed to be "led"?

File: samples/net/openthread/coap_server/src/coap_server.c [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+{
+	static u8_t val = 1;
+
+	dk_set_led(PROVISIONING_LED, val);
+	val = !val;
+}
+
+static void on_let_timer_stop(struct k_timer *timer_id)
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/coap_server.c]: typo

File: samples/net/openthread/coap_server/src/coap_server.c [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+static void on_provisioning_timer_expiry(struct k_timer *timer_id)
+{
+	deactivate_provisionig();
+}
+
+static void on_let_timer_expiry(struct k_timer *timer_id)
+{
+	static u8_t val = 1;
+
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/coap_server.c:81]: ARG_UNUSED

File: samples/net/openthread/coap_server/src/coap_server.c [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	dk_set_led(PROVISIONING_LED, val);
+	val = !val;
+}
+
+static void on_let_timer_stop(struct k_timer *timer_id)
+{
+	dk_set_led_off(PROVISIONING_LED);
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/coap_server.c:92]: ARG_UNUSED

File: samples/net/openthread/coap_server/src/ot_coap_utils.h [hunk 13]
```diff
@@ -0,0 +1,23 @@
+#ifndef __OT_COAP_UTILS_H__
+#define __OT_COAP_UTILS_H__
+
+#include <coap_server_client_interface.h>
+
+#include <openthread/instance.h>
+
+/**@brief Type definition of the function used to handle light resource change.
+ */
+typedef void (*light_request_callback)(light_command cmd);
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.h]: _t?

File: samples/net/openthread/coap_server/src/ot_coap_utils.c [hunk 14]
```diff
@@ -0,0 +1,247 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ot_coap_utils, CONFIG_OT_COAP_UTILS_LOG_LEVEL);
+
+#include "ot_coap_utils.h"
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.c]: nit: move this below zephyr includes

File: samples/net/openthread/coap_server/src/ot_coap_utils.c [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+
+#include <net/net_pkt.h>
+#include <net/net_l2.h>
+#include <net/openthread.h>
+#include <openthread/message.h>
+#include <openthread/ip6.h>
+#include <openthread/coap.h>
+#include <openthread/thread.h>
+
+typedef struct {
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.c]: doesn't it generate a warning? Not sure if its reported by compile or checkpatch but if I recall correctly typedefs are not allowed in c files.

File: samples/net/openthread/coap_server/src/ot_coap_utils.c [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+	struct otInstance *ot;
+	bool provisioning_enabled;
+	light_request_callback on_light_request;
+	provisioning_request_callback on_provisioning_request;
+} server_context_t;
+
+static server_context_t server_context = { .ot = NULL,
+					   .provisioning_enabled = false,
+					   .on_light_request = NULL,
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.c]: please double check it, but I think there was a warning about null initialized static variables.

File: samples/net/openthread/coap_server/src/ot_coap_utils.c [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+	struct net_if *iface;
+	struct openthread_context *ot_context;
+
+	iface = net_if_get_first_by_type(&NET_L2_GET_NAME(OPENTHREAD));
+	if (!iface) {
+		LOG_ERR("There is no net interface for OpenThread");
+		goto end;
+	}
+
+	ot_context = net_if_l2_data(iface);
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.c:51]: Is casting not needed here?

File: samples/net/openthread/coap_server/src/ot_coap_utils.c [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	error = otCoapStart(server_context.ot, COAP_PORT);
+	if (error != OT_ERROR_NONE) {
+		LOG_ERR("Failed to start OT CoAP. Error: %d", error);
+		goto end;
+	}
+
+end:
+	return error;
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.c]: return type is int, but error is of type otError. Need to change return type or error type. Not sure if returning otError in non-openthread code is a good idea.

**Review by MarekPorwisz — COMMENTED** [review]
left some minor comments.
I don't think we need workq for provisioning start as we only start two timers there but doing it that way doesn't hurt us so didn't left comment there.

File: samples/net/openthread/coap_server/src/ot_coap_utils.c (hunk 15 above)
Comment (lmaciejonczyk) [reply samples/net/openthread/coap_server/src/ot_coap_utils.c]: I don't see any warnings

File: samples/net/openthread/coap_server/src/ot_coap_utils.c (hunk 16 above)
Comment (lmaciejonczyk) [reply samples/net/openthread/coap_server/src/ot_coap_utils.c]: I don't see any warnings

**Comment by lmaciejonczyk** [comment]
> left some minor comments.

I've applied requested changes.

File: samples/net/openthread/coap_server/src/ot_coap_utils.c (hunk 17 above)
Comment (lmaciejonczyk) [reply samples/net/openthread/coap_server/src/ot_coap_utils.c:51]: > Is casting not needed here?

in C casting from void pointer is not needed

File: samples/net/openthread/coap_server/src/ot_coap_utils.h [hunk 19]
```diff
@@ -0,0 +1,23 @@
+#ifndef __OT_COAP_UTILS_H__
+#define __OT_COAP_UTILS_H__
+
+#include <openthread/instance.h>
+
+#include <coap_server_client_interface.h>
+
+/**@brief Type definition of the function used to handle light resource change.
+ */
+typedef void (*light_request_callback)(u8_t cmd);
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.h]: _t? (and line below), but can be left for now, until more changes are needed

File: samples/net/openthread/coap_server/src/ot_coap_utils.c [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	error = otCoapStart(srv_context.ot, COAP_PORT);
+	if (error != OT_ERROR_NONE) {
+		LOG_ERR("Failed to start OT CoAP. Error: %d", error);
+		goto end;
+	}
+
+end:
+	return error == OT_ERROR_NONE ? 0 : 1;
```
Comment (MarekPorwisz) [inline samples/net/openthread/coap_server/src/ot_coap_utils.c:248]: In zephyr errors are negative errno values eg -ENODEV, please use proper errno value or at least -1.

Comment (lmaciejonczyk) [reply samples/net/openthread/coap_server/src/ot_coap_utils.c:248]: this is application part of code not zephyr so that we should not stick to zephyr errno values

**Comment by lmaciejonczyk** [comment]
@rlubos Will you have time to look at this PR or we can merge it without your review?
